src/scripts/main.js

```diff
@@ -1,3 +1,418 @@
+const directions = {
+  ArrowUp: 'ArrowUp',
+  ArrowDown: 'ArrowDown',
+  ArrowLeft: 'ArrowLeft',
+  ArrowRight: 'ArrowRight',
+};
+const finalResult = 2048;
+
+startButton.addEventListener('click', () => {
+  if (startButton.classList.contains('start')) {
+    startButton.classList.remove('start');
+    startButton.classList.add('restart');
+    startButton.textContent = 'Restart';
+    startMessage.classList.add('hidden');
+
+    [...gameField].forEach((field, i) => {
+      fields[i] = field.innerText;
+    });
+  } else if (startButton.classList.contains('restart')) {
+    gameField.forEach(field => {
+      field.innerText = '';
+      field.classList.remove(...colorClasses);
+    });
+
+    for (let i = 0; i < 16; i++) {
+      fields[i] = '';
+    }
+    score.innerHTML = 0;
+    loseMessage.classList.add('hidden');
+    winMessage.classList.add('hidden');
+  }
+});
+
+document.addEventListener('keydown', event => {
+  if (!startButton.classList.contains('start')
+    && loseMessage.classList.contains('hidden')
+    && winMessage.classList.contains('hidden')) {
+    switch (event.key) {
+      case directions.ArrowUp:
+        upArrow(fields);
+        rematch(fields);
+        break;
+      case directions.ArrowDown:
+        downArrow(fields);
+        rematch(fields);
+        break;
+      case directions.ArrowLeft:
+        leftArrow(fields);
+        rematch(fields);
+        break;
+      case directions.ArrowRight:
+        rightArrow(fields);
+        rematch(fields);
+        break;
+      default:
+        return;
+    }
+
+    let freeFields = Object
+      .keys(Object
+        .fromEntries(Object
+          .entries(fields)
+          .filter(x => x[1] === '')));
+
+    const randomField = freeFields.random();
+    const addedNumber = Math.random() >= 0.1
+      ? 2
+      : 4;
+
+    fields[randomField] = addedNumber;
+    gameField[+randomField].textContent = addedNumber;
+
+    freeFields = Object
+      .keys(Object
+        .fromEntries(Object
+          .entries(fields)
+          .filter(x => x[1] === '')));
+
+    paint(gameField);
+
+    score.innerHTML = Object.values(fields).reduce((a, b) => +a + +b);
+
+    const isLose = lose(fields);
+
+    if (checkWin(fields)) {
+      winMessage.classList.remove('hidden');
+    }
+
+    if (freeFields.length === 0 && isLose) {
+      loseMessage.classList.remove('hidden');
+    }
+  }
+});
+
+const lose = (fields) => {
+  const endArray = [3, 7, 11, 15];
+  let counter = 0;
+
+  for (let i = 0; i <= 15; i++) {
+    if (
+      ((fields[i] !== fields[i + 1] && !endArray.includes(i))
+        || (endArray.includes(i) && fields[i] !== fields[i - 1]))
+      && fields[i] !== fields[i + 4]) {
+      counter++;
+    }
+  }
+
+  return counter === 16;
+};
+
+const checkWin = (object) => {
+  const values = Object.values(object);
+
+  return values.some(value => value === finalResult);
+};
+
+const upArrow = (fields) => {
+  for (let i = 4; i <= 7; i++) {
```

Comment:
There's a lot of repetitive code for handling arrow key movements. Consider creating a more generic function to handle all cases.